File: modules/web/js/ballerina/model/tree/tree/abstract-tree/annotation-node.js
/**
 * Copyright (c) 2017, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
 *
 * WSO2 Inc. licenses this file to you under the Apache License,
 * Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import Node from '../node';

class AnnotationNodeAbstract extends Node {


    setName(newValue, title) {
        let oldValue = this.name;
        title = (_.isNil(title))? 'Modify ${child.kind}':title;
        this.name = newValue;
        this.trigger('tree-modified', {
            origin: this,
            type: 'modify-node',
            title,
            data: {
                attributeName: 'name',
                newValue,
                oldValue,
            }
        });
    }

    getName() {
        return this.name;
    }



    setAttributes(newValue, title) {
        let oldValue = this.attributes;
        title = (_.isNil(title))? 'Modify ${child.kind}':title;
        this.attributes = newValue;
        this.trigger('tree-modified', {
            origin: this,
            type: 'modify-node',
            title,
            data: {
                attributeName: 'attributes',
                newValue,
                oldValue,
            }
        });
    }

    getAttributes() {
        return this.attributes;
    }


    addAttributes(node, i = -1){
        node.parent = this;
        let index = i;
        if (i === -1) {
            this.attributes.push(node);
            index = this.attributes.length;
        } else {
            this.attributes.splice(i, 0, node);
        }
        this.trigger('tree-modified', {
            origin: this,
            type: 'child-added',
            title: `Add ${child.kind}`,
            data: {
                child,
                index,
            },
        });
    }


    setFlags(newValue, title) {
        let oldValue = this.flags;
        title = (_.isNil(title))? 'Modify ${child.kind}':title;
        this.flags = newValue;
        this.trigger('tree-modified', {
            origin: this,
            type: 'modify-node',
            title,
            data: {
                attributeName: 'flags',
                newValue,
                oldValue,
            }
        });
    }

    getFlags() {
        return this.flags;
    }



    setAnnotationAttachments(newValue, title) {
        let oldValue = this.annotationAttachments;
        title = (_.isNil(title))? 'Modify ${child.kind}':title;
        this.annotationAttachments = newValue;
        this.trigger('tree-modified', {
            origin: this,
            type: 'modify-node',
            title,
            data: {
                attributeName: 'annotationAttachments',
                newValue,
                oldValue,
            }
        });
    }

    getAnnotationAttachments() {
        return this.annotationAttachments;
    }


    addAnnotationAttachments(node, i = -1){
        node.parent = this;
        let index = i;
        if (i === -1) {
            this.annotationAttachments.push(node);
            index = this.annotationAttachments.length;
        } else {
            this.annotationAttachments.splice(i, 0, node);
        }
        this.trigger('tree-modified', {
            origin: this,
            type: 'child-added',
            title: `Add ${child.kind}`,
            data: {
                child,
                index,
            },
        });
    }


    setWS(newValue, title) {
        let oldValue = this.wS;
        title = (_.isNil(title))? 'Modify ${child.kind}':title;
        this.wS = newValue;
        this.trigger('tree-modified', {
            origin: this,
            type: 'modify-node',
            title,
            data: {
                attributeName: 'wS',
                newValue,
                oldValue,
            }
        });
    }

    getWS() {
        return this.wS;
    }



    setKind(newValue, title) {
        let oldValue = this.kind;
        title = (_.isNil(title))? 'Modify ${child.kind}':title;
        this.kind = newValue;
        this.trigger('tree-modified', {
            origin: this,
            type: 'modify-node',
            title,
            data: {
                attributeName: 'kind',
                newValue,
                oldValue,
            }
        });
    }

    getKind() {
        return this.kind;
    }



    setPosition(newValue, title) {
        let oldValue = this.position;
        title = (_.isNil(title))? 'Modify ${child.kind}':title;
        this.position = newValue;
        this.trigger('tree-modified', {
            origin: this,
            type: 'modify-node',
            title,
            data: {
                attributeName: 'position',
                newValue,
                oldValue,
            }
        });
    }

    getPosition() {
        return this.position;
    }



}

export default AnnotationNodeAbstract;

File: compiler/ballerina-parser/src/main/java/io/ballerina/compiler/syntax/tree/AnnotationNode.java
/*
 *  Copyright (c) 2020, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
 *
 *  WSO2 Inc. licenses this file to you under the Apache License,
 *  Version 2.0 (the "License"); you may not use this file except
 *  in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing,
 *  software distributed under the License is distributed on an
 *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *  KIND, either express or implied.  See the License for the
 *  specific language governing permissions and limitations
 *  under the License.
 */
package io.ballerina.compiler.syntax.tree;

import io.ballerina.compiler.internal.parser.tree.STNode;

import java.util.Objects;
import java.util.Optional;

/**
 * This is a generated syntax tree node.
 *
 * @since 2.0.0
 */
public class AnnotationNode extends NonTerminalNode {

    public AnnotationNode(STNode internalNode, int position, NonTerminalNode parent) {
        super(internalNode, position, parent);
    }

    public Token atToken() {
        return childInBucket(0);
    }

    public Node annotReference() {
        return childInBucket(1);
    }

    public Optional<MappingConstructorExpressionNode> annotValue() {
        return optionalChildInBucket(2);
    }

    @Override
    public void accept(NodeVisitor visitor) {
        visitor.visit(this);
    }

    @Override
    public <T> T apply(NodeTransformer<T> visitor) {
        return visitor.transform(this);
    }

    @Override
    protected String[] childNames() {
        return new String[]{
                "atToken",
                "annotReference",
                "annotValue"};
    }

    public AnnotationNode modify(
            Token atToken,
            Node annotReference,
            MappingConstructorExpressionNode annotValue) {
        if (checkForReferenceEquality(
                atToken,
                annotReference,
                annotValue)) {
            return this;
        }

        return NodeFactory.createAnnotationNode(
                atToken,
                annotReference,
                annotValue);
    }

    public AnnotationNodeModifier modify() {
        return new AnnotationNodeModifier(this);
    }

    /**
     * This is a generated tree node modifier utility.
     *
     * @since 2.0.0
     */
    public static class AnnotationNodeModifier {
        private final AnnotationNode oldNode;
        private Token atToken;
        private Node annotReference;
        private MappingConstructorExpressionNode annotValue;

        public AnnotationNodeModifier(AnnotationNode oldNode) {
            this.oldNode = oldNode;
            this.atToken = oldNode.atToken();
            this.annotReference = oldNode.annotReference();
            this.annotValue = oldNode.annotValue().orElse(null);
        }

        public AnnotationNodeModifier withAtToken(
                Token atToken) {
            Objects.requireNonNull(atToken, "atToken must not be null");
            this.atToken = atToken;
            return this;
        }

        public AnnotationNodeModifier withAnnotReference(
                Node annotReference) {
            Objects.requireNonNull(annotReference, "annotReference must not be null");
            this.annotReference = annotReference;
            return this;
        }

        public AnnotationNodeModifier withAnnotValue(
                MappingConstructorExpressionNode annotValue) {
            Objects.requireNonNull(annotValue, "annotValue must not be null");
            this.annotValue = annotValue;
            return this;
        }

        public AnnotationNode apply() {
            return oldNode.modify(
                    atToken,
                    annotReference,
                    annotValue);
        }
    }
}
